interpreter/operator_string.go

```diff
@@ -355,4 +355,82 @@ func evalStartsWith(m model.IBinaryExpression, lObj, rObj result.Value) (result.
 		return result.Value{}, err
 	}
 	return result.New(strings.HasPrefix(argument, prefix))
-}
\ No newline at end of file
+}
+
+// Substring(stringToSub String, startIndex Integer) String
+// Substring(stringToSub String, startIndex Integer, length Integer) String
+// https://cql.hl7.org/09-b-cqlreference.html#substring
+func evalSubstring(m model.INaryExpression, operands []result.Value) (result.Value, error) {
+	if len(operands) < 2 || len(operands) > 3 {
+		return result.Value{}, fmt.Errorf("Substring expects 2 or 3 arguments, got %d", len(operands))
+	}
+
+	// Operand 0: stringToSub (String)
+	if result.IsNull(operands[0]) {
+		return result.New(nil)
+	}
+	stringToSub, err := result.ToString(operands[0])
+	if err != nil {
+		return result.Value{}, fmt.Errorf("could not convert stringToSub to string: %w", err)
+	}
+
+	// Operand 1: startIndex (Integer)
+	if result.IsNull(operands[1]) {
+		return result.New(nil)
+	}
+	startIndex, err := result.ToInt32(operands[1])
+	if err != nil {
+		return result.Value{}, fmt.Errorf("could not convert startIndex to int32: %w", err)
+	}
+
+	runes := []rune(stringToSub)
+	stringLen := int32(len(runes))
+
+	// Rule: If the stringToSub is null or the startIndex is null, the result is null.
+	// (Already handled by initial checks on operands[0] and operands[1])
+
+	// Special case: If string is empty and startIndex is 0, return empty string
+	if stringLen == 0 && startIndex == 0 {
+		return result.New("")
+	}
```

Comment:
what should Substring('', 0) produce? The current code returns an empty string instead of Null (it seems reasonable to get the identical string when doing Substring(str, 0) == str to me). The spec would probably imply Null. Could be worth clarifying/discussing, happy to change that back.